src/frequenz/sdk/actor/_background_service.py

```diff
@@ -0,0 +1,307 @@
+
+        async def main() -> None:
+            # As an async context manager
+            async with Clock(resolution_s=1):
+                await asyncio.sleep(5)
+
+            # Manual start/stop (only use if necessary, as cleanup is more complicated)
+            clock = Clock(resolution_s=1)
+            await clock.start()
+            await asyncio.sleep(5)
+            await clock.stop()
+
+        asyncio.run(main())
+        ```
+
+    Example: Complete example
+        This is the same example as the *Simple example*, but allowing users to specify
+        a custom loop (and name). This option should be preferred when writing
+        background services, to give users full flexibility.
+
+        ```python
+        import datetime
+        import asyncio
+
+
+        class Clock(BackgroundService):
+            def __init__(
+                self,
+                resolution_s: float,
+                *,
+                name: str | None = None,
+                loop: asyncio.AbstractEventLoop | None = None,
+            ) -> None:
+                super().__init__(name=name, loop=loop)
+                self._resolution_s = resolution_s
+
+            async def start(self) -> None:
+                # Note that you should always use the self._loop attribute access loop
+                # functions instead of asyncio directly.
+                self._tasks.add(self.loop.create_task(self._tick()))
+
+            async def _tick(self) -> None:
+                while True:
+                    await asyncio.sleep(self._resolution_s)
+                    print(datetime.datetime.now())
+
+
+        async def main(loop: asyncio.AbstractEventLoop) -> None:
+            print("Clock as async context manager")
+            async with Clock(resolution_s=1, name="Clock with context", loop=loop):
+                await asyncio.sleep(5)
+
+            print("Clock with manual start/stop")
+            clock = Clock(resolution_s=1, name="Manual clock", loop=loop)
+            await clock.start()
+            await asyncio.sleep(5)
+            await clock.stop()
+
+
+        main_loop = asyncio.get_event_loop_policy().new_event_loop()
+        main_loop.run_until_complete(main(main_loop))
+        ```
+    """
+
+    def __init__(
+        self, *, name: str | None = None, loop: asyncio.AbstractEventLoop | None = None
+    ) -> None:
+        """Initialize this BackgroundService.
+
+        Args:
+            name: The name of this background service. If `None`, `str(id(self))` will
+                be used. This is used mostly for debugging purposes.
+            loop: The event loop this background service will run on. If `None`, the
+                current running event loop will be used.
+        """
+        self._name: str = str(id(self)) if name is None else name
+        self._loop: asyncio.AbstractEventLoop = (
+            loop if loop else asyncio.get_running_loop()
+        )
+        self._tasks: set[asyncio.Task[Any]] = set()
```

Comment:
I wonder if `self._name` and `self._tasks` can be **also** initialized in the `start` method,  so that people don't have to call `super.__init__()` from actor init methods,  even if they don't want to set a name.

Reply:
I don't understand this. You mean initialize it if it wasn't set in the constructor?

In any case I think it will make things more complicated, for example the `run()` function emits some logs that actors will be started before they are started, so in that case they won't have the names set. Also we'll need to make `_name` and `_tasks` optional (`type | None`, which will force us to litter the implementation with `assert _ not None`). I think having to write the `super().__init__()` is not that bad, also some of the linters (I don't remember if it is `mypy` or `pylint`) catches the issue it if you forget.

Reply:
> You mean initialize it if it wasn't set in the constructor?

yup,  but if pylint, etc is able to catch a missing `super().__init__()` that is just as good.  Is there a test for this case already?

Reply:
You mean a test that `pylint` catches it?

Reply:
yup

Reply:
Oh,  but pylint doesn't complain if there are unused `disable=` in the code.  Unlike mypy which complains if there are unused `type: ignore`

Reply:
Yeah, I'm not sure how I would be able to test if pylint keeps catching this 😕

Reply:
BTW, it is pylint that catches this: https://pylint.readthedocs.io/en/latest/user_guide/messages/warning/super-init-not-called.html